### DISCUSSION TITLE
Bootstrap 5 RTL

### DISCUSSION THREAD

**Discussion by islamx** [twbs/bootstrap#35992] [Q&A]
Hello, i am asking about RTL scss in bootstrap 5, seems the documentation misleading and i am feeling lost a bit
i am trying to modify the utitlits to be for ex
`
LTR
.ms-1 >> margin-left
`
i need to be in RTL
`
ms-1 >> margin-right
`
as the start for RTL is right not left, and this for the rest of alignment 
text-direction, float, borders

any help?

**Comment by florianlacreuse** [comment]
Hi,

Did you read this part of the docs: https://getbootstrap.com/docs/5.1/getting-started/rtl/ ?

**Reply by islamx** [reply]
yes, but there is something i am not understand  to find answer to my question

**Comment by motahharm** [comment]
Hi @islamx 

your html tag must have dir attr with rtl value like this:
`<html dir="rtl">`
or you can use css:
`html {direction: rtl;}`

or you can use <a href="https://rtlcss.com/">RTLcss</a>.

**Reply by islamx** [reply]
i did this and the whole layout render well but still i have a problem with the alignment like margins and padding and borders

in my current project that i have
i use ms-1 for ex in RTL direction and this return margin **left** in spite of it should return margin **right** not left

So from where should i set the start & end for each direction?

**Reply by motahharm** [reply]
Hi
You should add `dir` property with `rtl` value to your `html` tag or any other root elements like `div`.
Then you can use `ms-4` or any other classes that work with `start` or `end` properties.

**Comment by islamx** [comment]
other thing that i got
after i import 
`<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.rtl.min.css" integrity="sha384-gXt9imSW0VcJVHezoNQsP+TNrjYXoGcrqBZJpry9zJt8PCQjobwmhMGaDHTASo9N" crossorigin="anonymous">`
everything be fine
but there is something happen that i am not got it

I've add custom-variables to override the primary, secondary colors to adjust it based on the brand that i code it
the the file doesn't see the custom variables 

and here is what i did

in index.html add in head
`<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.rtl.min.css" integrity="sha384-gXt9imSW0VcJVHezoNQsP+TNrjYXoGcrqBZJpry9zJt8PCQjobwmhMGaDHTASo9N" crossorigin="anonymous">`

and in html tag i've added 
`lang="ar" dir="rtl"`

and in app.scss i've added custom.scss then in this file

my custom variables like

`$primary: pink;`

then i called
`@import "~bootstrap/scss/functions";
@import "~bootstrap/scss/variables";
@import "~bootstrap/scss/mixins";`

and the file doesn't respect my new variable

**Reply by ayamoh001** [reply]
In the docs, they said that I must import functions and variables first then modify colors then import the rest of bootstrap or what you need from it.

and I have a similar issue, I wanna to complie my custimized sass file to get both ltr and rtl versions.

**Reply by sulavalentin** [reply]
did you resolve your problem? i have the same issue

**Comment by ibrahimAboelsuod** [comment]
What I think is missing from the documents is the fact that **if you are importing the scss version**, then you need to have postcss installed and configured. AND you also need an rtl plugin. I had success with [postcss-rtlcss](https://github.com/elchininet/postcss-rtlcss).

So just install postcss & postcss-rtlcss:

`npm i -D postcss postcss-rtlcss`

Then in your project root (or in the app dir if you are using nx) add this config file named `postcss.config.js`:

```
module.exports = {
  plugins: ['postcss-rtlcss']
};
```
And now you just run your project and should have rtl working perfectly!

**Reply by omaraalia93** [reply]
not working

**Reply by ibrahimAboelsuod** [reply]
> not working 

How are you importing bootstrap?

**Reply by omaraalia93** [reply]
In the global "styles.scss" file, I imported the Bootstrap SCSS file as follows:
@import "../node_modules/bootstrap/scss/bootstrap.scss";

Afterward, I followed the steps you outlined. However, I noticed that the "ms-2" class from "_utilities.scss" still results in a margin-left on both directions.

**Reply by omaraalia93** [reply]
any update please?

**Reply by ibrahimAboelsuod** [reply]
I had this exact problem before. I am trying to remember how I solved it. Will update you once I do.

**Reply by ibrahimAboelsuod** [reply]
@omaraalia93 I think had this problem when I was importing both normal bootstrap and the rtl version by mistake. 

Do you have `dir="rtl"` and `lang="ar"` on your html tag?

Also could you inspect the element with `ms-2` and see where the css rules are coming from?

**Reply by omaraalia93** [reply]
yes i have dir="rtl" and lang="ar" attributes.

webpack:///node_modules/bootstrap/scss/mixins/_utilities.scss

.ms-2 {
margin-left: 0.5rem !important;
}

**Reply by ibrahimAboelsuod** [reply]
I am not sure what could be wrong.

But check these 2 links. [Postcss docs](https://github.com/postcss/postcss) check the webpack section maybe there is something missing - I think you are using webpack.

Also there this next + postcss-rtlcss [example project](https://github.com/elchininet/nextjs-postcss-rtlcss-example) which gonna be different from your setup but it might help.

I am very curious what is causing this so please update me of your findings.

**Reply by omaraalia93** [reply]
ok i will and thanks.

**Reply by omaraalia93** [reply]
do you have any working example for an angular project with bootstrap v5 implemented properly?

**Reply by ibrahimAboelsuod** [reply]
@omaraalia93 I have one in nextjs. All my angular projects are still in bootstrap v4 unfortunately. 

If you wanna email me we can get on a call and maybe I can help you figure it out. here is my email ibrahimaboelsuod@gmail.com

**Reply by Filini** [reply]
This is how I did it in Angular@16, with SASS and RTLCSS in package.json, before building the application:
1. build Bootstrap SCSS (with your customized variables), with sass, as /assets/bootstrap.min.css
2. build the RTL file with rtlcss, from /assets/bootstrap.min.css to /assets/bootstrap.rtl.min.css
3. in your index.html. link the /assets/bootstrap.rtl.min.css file, and set dir="rtl" and lang="ar" in the HTML element

So, my build script in package.json looks something like:
"build-all": "sass ... bootstrap.scss ./assets/bootstrap.min.css & rtlcss ... ./assets/bootstrap.min.css ./assets/bootstrap.rtl.min.css & ng build ...",
(the script paths depend on the bootstrap files location, and root options in angular.json, but you get the idea)

If you need to switch between LTR and RTL without reloading the page, you can update the DOM to change the link SRC and the dir/lang attributes

**Comment by 3baaady07** [comment]
As per the [Bootstrap docs](https://getbootstrap.com/docs/5.3/getting-started/rtl/#get-familiar), you need a tool ([`rtlcss`](https://rtlcss.com/)) to process the generated css file from your scss project. Then `rtlcss`, in turn, generates an RTL css file.

1. Step One: compile sass to css (eg. `style.css`)
2. Step two: run `rtlcss` on `style.css` and generate `style.rtl.css`
3. Step three: manually change the stylesheet file on the document using javascript based on event

**Comment by PrabagaranSaminathan** [comment]
@islamx, Do you get any updates?. I also facing same issue

**Reply by sulavalentin** [reply]
i too, did you find the solution?

**Reply by PrabagaranSaminathan** [reply]
@sulavalentin,  I have created a custom Bootstrap RTL SCSS and written the necessary CSS in that SCSS

 [dir='rtl'] {
   @import './bootstrap-rtl';
 }

**Reply by elchininet** [reply]
@PrabagaranSaminathan,
Please, [check this discussion](https://github.com/orgs/twbs/discussions/39726).
That approach has many drawbacks.
Regards

**Comment by ANU1o** [comment]
Some bootstrap classes are still undercooked when it comes to `rtl`. 

i use `mx` tag instead of `ms` or `me` when it comes to multi language sites (that has both `ltr` and `rtl`). To a degree, it gives symmetry.

And yes default `ms` and `me` seems misleading when doing website in `rtl` format.

for now i use custom css like `[dir="rtl"] .ms-<X> {margin-right: <valueBasedonX>}`. Extra work but helps in avoiding confusion.
